iocage/lib/Jail.py

```diff
@@ -0,0 +1,500 @@
+            raise Exception(msg)
+
+    def require_jail_stopped(self):
+        if self.running:
+            msg = f"Jail {self.humanreadable_name} is already running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def require_jail_running(self):
+        if not self.running:
+            msg = f"Jail {self.humanreadable_name} is not running"
+            self.logger.error(msg)
+            raise Exception(msg)
+
+    def update_jail_state(self):
+        try:
+            stdout = subprocess.check_output([
+                "/usr/sbin/jls",
+                "-j",
+                self.identifier,
+                "-v",
+                "-h"
+            ], shell=False, stderr=subprocess.DEVNULL)
+            output = stdout.decode("utf-8").strip()
+
+            keys, values = [x.split(" ") for x in output.split("\n")]
+            self.jail_state = dict(zip(keys, values))
+
+        except:
+            self.jail_state = None
+
+    def _resolve_uuid(self, text):
+        jails_dataset = self.host.datasets.jails
+        for dataset in list(jails_dataset.children):
+            dataset_uuid = dataset.name[(len(jails_dataset.name) + 1):]
+            if text in dataset_uuid:
+                self.logger.debug(f"Resolved {text} to uuid {dataset_uuid}")
+                return dataset_uuid
+
+        raise Exception(f"No jail matching {text} was found")
+
+    def _get_name(self):
+        return self.humanreadable_name
+
+    def _get_humanreadable_name(self):
+
+        try:
+            return self.config.name
+        except:
+            pass
+
+        try:
+            return str(self.config.uuid)[:8]
+        except:
+            pass
+
+        raise Exception("This Jail does not have any identifier yet")
+
+    def _get_stopped(self):
+        return self.running is not True
+
+    def _get_running(self):
+        return self._get_jid() is not None
+
+    def _get_jid(self):
+        try:
+            return self.jail_state["jid"]
+        except:
+            pass
+
+        try:
+            self.update_jail_state()
+            return self.jail_state["jid"]
+        except:
+            return None
+
+    def _get_identifier(self):
+        return f"ioc-{self.uuid}"
+
+    def _get_exists(self):
+        try:
+            self.dataset
+            return True
+        except:
+            return False
+
+    def _get_uuid(self):
+        return self.config.uuid
+
+    def _get_release(self):
+        return Release.Release(
+            name=self.config.release,
+            logger=self.logger,
+            host=self.host,
+            zfs=self.zfs
+        )
+
+    def _get_jail_type(self):
+        return self.config.type
+
+    def _get_dataset_name(self):
+        return f"{self.host.datasets.root.name}/jails/{self.config.uuid}"
+
+    def _get_dataset(self):
+        return self.zfs.get_dataset(self._get_dataset_name())
+
+    def _get_path(self):
+        return self.dataset.mountpoint
```

Comment:
`_get_path` is a private function, so it would be up to the class to call it only after initializing a dataset. When the dataset does not exist, it needs to be created before checking the path. In this case even the private function is never called and will be entirely removed.